### PR TITLE
Handle broken legacy redirects with 404 

### PR DESCRIPTION
**Pull request by kerwinso**
For bug #452. Tested on my local machine.

### CODE REVIEW

**Comment by kerwinso** [comment]
Travis tests are failing on master, not related what I updated (I think). See #478.

File: cnxarchive/views.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ def is_latest(id, version):
 
 
 TYPE_INFO = []
-LEGACY_EXTENSION_MAP = {'epub': 'epub', 'pdf': 'pdf', 'zip': 'complete.zip'}
+LEGACY_EXTENSION_MAP = {'epub': 'epub', 'pdf': 'pdf', 'zip': 'offline.zip'}
```
Comment (reedstrm) [inline cnxarchive/views.py]: I though we consciously made the choice to serve the complete, not the offline, since the offline does not actually work for what it claims to be (viewable offline) - in fact it fails to build in some cases, and is built _from_ the complete zip, so complete zip should exist wherever offline zip does. Is there a bug number that references this?

File: cnxarchive/views.py [hunk 2]
```diff
@@ -260,13 +260,16 @@ def _get_page_in_book(page_uuid, page_version, book_uuid,
...
-    pages = list(flatten_tree_to_ident_hashes(coltree))
-    page_ident_hash = join_ident_hash(page_uuid, page_version)
-    if page_ident_hash in pages:
-        return book_uuid, '{}:{}'.format(
-            latest and book_uuid or book_ident_hash, page_uuid)
-    # book not in page
-    return page_uuid, page_ident_hash
+    if coltree is None:
+        raise httpexceptions.HTTPNotFound()
+    else:
```
Comment (reedstrm) [inline cnxarchive/views.py]: while this makes the logic clear,  raising the exception above aborts execution, so you can avoid the extra indent by not using the `else`. It's a style choice though, this is fine.

File: cnxarchive/views.py [hunk 3]
```diff
@@ -260,13 +260,16 @@ def _get_page_in_book(page_uuid, page_version, book_uuid,
...
     coltree = _get_content_json(ident_hash=book_ident_hash)['tree']
-    pages = list(flatten_tree_to_ident_hashes(coltree))
-    page_ident_hash = join_ident_hash(page_uuid, page_version)
-    if page_ident_hash in pages:
-        return book_uuid, '{}:{}'.format(
-            latest and book_uuid or book_ident_hash, page_uuid)
-    # book not in page
-    return page_uuid, page_ident_hash
+    if coltree is None:
+        raise httpexceptions.HTTPNotFound()
```
Comment (reedstrm) [inline cnxarchive/views.py]: no coverage on the exception. Could use a test that attempts to retrieve a non existent uuid. See my push.

File: cnxarchive/views.py [hunk 4]
```diff
@@ -260,13 +260,16 @@ def _get_page_in_book(page_uuid, page_version, book_uuid,
...
+    else:
+        pages = list(flatten_tree_to_ident_hashes(coltree))
+        page_ident_hash = join_ident_hash(page_uuid, page_version)
+        if page_ident_hash in pages:
+            return book_uuid, '{}:{}'.format(
+                latest and book_uuid or book_ident_hash, page_uuid)
+        # book not in page
+        return page_uuid, page_ident_hash
```
Comment (reedstrm) [inline cnxarchive/views.py]: also not covered, but not your fault. We could use a this page not in this book test as well.

File: cnxarchive/views.py (hunk 3 above)
Comment (reedstrm) [inline cnxarchive/views.py]: ab96c10fc5c1ffea88f7987108f40954d11b2a43

File: cnxarchive/views.py (hunk 1 above)
Comment (kerwinso) [inline cnxarchive/views.py]: See Karen's comment in #420, which was the basis for this change. But based on your comment, it sounds like this change should be reverted.

File: cnxarchive/views.py (hunk 2 above)
Comment (kerwinso) [inline cnxarchive/views.py]: I actually wasn't aware of that, though it makes sense, so thank you. :themoreyouknow:

File: cnxarchive/views.py (hunk 3 above)
Comment (kerwinso) [inline cnxarchive/views.py]: Thank you for writing the test. As a n00b I can't confirm whether it looks good or not. @pumazi and I started discussing writing unit tests before his vacation. Now it's just a matter of getting some of his time for another pair programming session.

File: cnxarchive/tests/test_views.py [hunk 5]
```diff
@@ -1444,6 +1444,20 @@ def test_exports_without_version(self):
...
+        version = '1.1'
+
+        # Build the request
+        self.request.matchdict = {'ident_hash': '{}@{}'.format(id, version)}
+        self.request.matched_route = mock.Mock()
+        self.request.matched_route.name = 'content-extras'
+
+        from ..views import get_extra
+
+        self.assertRaises(httpexceptions.HTTPNotFound,
```
Comment (pumazi) [inline cnxarchive/tests/test_views.py]: @reedstrm For future reference, it's often times easiest to use the `with` statement when using the `assertRaises` because it makes it far easier to step into the function in question from a `pdb` prompt.

File: cnxarchive/views.py (hunk 2 above)
Comment (kerwinso) [inline cnxarchive/views.py]: Extra indent and `else` removed with https://github.com/Connexions/cnx-archive/pull/477/commits/e6743ba54efb8bb900947541bc3367835284a31b.

File: cnxarchive/views.py (hunk 1 above)
Comment (kerwinso) [inline cnxarchive/views.py]: I apologize for including the change for #420 in this PR. It should be handled in a separate PR. I reverted the change with https://github.com/Connexions/cnx-archive/pull/477/commits/e6743ba54efb8bb900947541bc3367835284a31b.

File: cnxarchive/views.py (hunk 3 above)
Comment (kerwinso) [inline cnxarchive/views.py]: New test written with @pumazi, included in this commit: https://github.com/Connexions/cnx-archive/pull/477/commits/e6743ba54efb8bb900947541bc3367835284a31b

**Comment by kerwinso** [comment]
I should not have included an attempted fix for #420 in this PR. This PR should only deal with bug #452. I've backed out the change related to #420 with commit https://github.com/Connexions/cnx-archive/pull/477/commits/e6743ba54efb8bb900947541bc3367835284a31b. 

I've also re-titled this PR to be more descriptive. Apologies for the confusion.

File: cnxarchive/tests/test_views.py [hunk 6]
```diff
@@ -1444,6 +1469,27 @@ def test_exports_without_version(self):
...
+        self.request.matched_route = mock.Mock()
+        self.request.matched_route.name = 'content-extras'
+
+        from ..views import get_extra
+
+        # self.assertRaises(httpexceptions.HTTPNotFound,
+        #                   get_extra, self.request)
+
+        with self.assertRaises(httpexceptions.HTTPNotFound) as caught_exc:
+            import pdb;
```
Comment (reedstrm) [inline cnxarchive/tests/test_views.py]: oops!

Comment (kerwinso) [inline cnxarchive/tests/test_views.py]: Addressed in https://github.com/Connexions/cnx-archive/pull/477/commits/4aa2b25d8b68607229d228d30fe195fd45aeb69d

File: cnxarchive/tests/test_views.py [hunk 7]
```diff
@@ -1444,6 +1469,25 @@ def test_exports_without_version(self):
         with self.assertRaises(IdentHashMissingVersion) as cm:
             get_export(self.request)
 
+    def test_get_extra_404(self):
+        #id = 'b771c6fc-34f0-11e7-ad77-e3343f783f02'
```
Comment (reedstrm) [inline cnxarchive/tests/test_views.py]: remove dead-code comments

File: cnxarchive/tests/test_views.py [hunk 8]
```diff
@@ -1444,6 +1469,25 @@ def test_exports_without_version(self):
...
+        version = '66.1'
+
+        # Build the request
+        self.request.matchdict = {'ident_hash': '{}@{}'.format(id, version)}
+        self.request.matched_route = mock.Mock()
+        self.request.matched_route.name = 'content-extras'
+
+        from ..views import get_extra
+
+        # self.assertRaises(httpexceptions.HTTPNotFound,
```
Comment (reedstrm) [inline cnxarchive/tests/test_views.py]: ditto

Comment (kerwinso) [inline cnxarchive/tests/test_views.py]: done

File: cnxarchive/tests/test_views.py (hunk 7 above)
Comment (kerwinso) [inline cnxarchive/tests/test_views.py]: done
